### PR TITLE
Add MDS Stops RT Specification (Agency & Provider)

### LINKED ISSUE #374: Docked Bike Share Systems
### Is your feature request related to a problem? Please describe.

I'd love to see the ability to support docked bike share systems in MDS, with the appropriate changes to reflect the lack of GPS / the introduction of docks in the system. For example, trips/ route info is probably impossible to get. 

Prior art: 

[LA metrobike historical trips](https://bikeshare.metro.net/about/data/)
[Motivate Bay Wheels System Data](https://www.lyft.com/bikes/bay-wheels/system-data) 
[BikeshareMap](https://bikesharemap.com/) 
[Statistical Models around bikeshare data / Divvy](https://github.com/dssg/bikeshare) 

### Describe the solution you'd like

MDS, but for docked bikeshare systems. 

### Is this a breaking change

A breaking change would require consumers or implementors of the API to modify their code for it to continue to function (ex: renaming of a required field or the change in data type of an existing field). A non-breaking change would allow existing code to continue to function (ex: addition of an optional field or the creation of a new optional endpoint). 

 * I'm not sure

### `Provider` or `agency`

For which API is this feature being requested:

 * `provider`

### Describe alternatives you've considered

Continue to get CSV dumps in different formats across different operators.

### PR DESCRIPTION
**Pull request by avatarneil**
### Explain pull request
This PR is a work in progress to resolve #374 .

### Is this a breaking change

A breaking change would require consumers or implementors of the API to modify their code for it to continue to function (ex: renaming of a required field or the change in data type of an existing field). A non-breaking change would allow existing code to continue to function (ex: addition of an optional field or the creation of a new optional endpoint). 

* No, not breaking

### Impacted Spec

Which spec(s) will this pull request impact?

 * `Provider`
 * `Agency`

### Additional context

`Stops` are intended to be a strict superset of GTFS/GBFS -- keep that in mind when seeing some fields that are unnecessary for the base scope of MDS currently.

### CODE REVIEW

**Comment by dirkdk** [comment]
quick question: how would stops with charging capabilities be reflected?

**Comment by avatarneil** [comment]
@thekaveman Resolved your previous nit. Also cleaned up the commit history a bit in the process.

**Comment by avatarneil** [comment]
@thekaveman I've also removed any `agency` changes that were included in this, and have cut a separate PR.

**Comment by dirkdk** [comment]
I guess this PR does not define vehicle_type yet. In order to support charging, we could define them as following with the `_charging` postfix:

```
[
num_bike_human_available, 
num_bike_electric_assist_available, 
num_bike_electric_assist_available_charging, 
num_scooter_electric_available, 
num_scooter_electric_available_charging
]
```

e.g.: 
```
{
  num_bike_human_available: 10,
  num_bike_electric_assist_available: 5,
  num_bike_electric_assist_available_charging: 3,
  num_scooter_electric_available: 3,
  num_scooter_electric_available_charging: 2
}
```

**Comment by avatarneil** [comment]
@dirkdk Vehicle Types are defined here https://github.com/openmobilityfoundation/mobility-data-specification/tree/dev/provider#vehicle-types already, I add a hyperlink to the line that references that enum.

**Comment by dirkdk** [comment]
> @dirkdk Vehicle Types are defined here https://github.com/openmobilityfoundation/mobility-data-specification/tree/dev/provider#vehicle-types already, I can add a hyperlink to the line that references that enum.

yes thinking if we need to include `propulsion_type`. Definitely we need to distinct non-charging racks from charging racks

**Comment by jiffyclub** [comment]
I'm not as familiar with GBFS so this might not be compatible with that, but instead of listing vehicle counts and having to invent increasingly complex keys for that I'd rather see a list of vehicles with associated attributes so end users can group and count those as needed.

**Comment by jfh01** [comment]
There are some relevant changes likely coming to GBFS here: https://github.com/NABSA/gbfs/pull/219

Worth making sure we have alignment with those (especially around valet and virtual stations) in whatever we implement.

**Comment by vperron** [comment]
Hi, I must say that I like this approach very much since:
- it's a pragmatic one that aims to support a reasonable level of information from the GBFS standard
- it's generic enough to be able to evolve and support dockless mobility in future versions (for instance it already supports various vehicle types and expected geography IDs for virtual stations)
- has a good chance to be integrated into the 1.0.0 milestone without too much hassle, especially if the changes to `/trips` and `/status_changes` are out of scope for 1.0 so far.

I'm not that sure about the `wheelchair_boarding` boolean though, I think it's be more generic to have a list of "available services" that we can extend, much like the `propulsion_types` for instance.

However, I think that https://github.com/openmobilityfoundation/mobility-data-specification/pull/441, while maybe going maybe too far for a first step in my opinion, also makes some very interesting points, most notably:
- it provides an optional `gbfs_station_id`: being able to reconcile MDS info and GBFS can be a very big plus for some peeps who would want to use both approaches.
- a `status` field that reports the whole station status: it happens _a lot_ that some stations are disabled, decommissioned, temporarily closed, ...
- a `reported_at` field: we need to be able to know, in a real-time endpoint, how stale this information can be. I think this is really important.

I also think that their approach with arrays of MDS-compatible vehicle types, propulsion types, vehicles and docks is cleaner and more informative than multiple dictionaries of vehicles and spaces counts, which sounds very  GBFS-y :)

So to sum it up I'd say that in my opinion, we can find a best of both worlds by merging some of the fields in that other PR with this one and still be generic and simple enough for a first approach.

**Comment by jfh01** [comment]
Here is a doc we put together with MobilityData and NABSA to provide additional guidance on how GBFS and MDS should align with each other:

https://github.com/openmobilityfoundation/governance/blob/master/technical/GBFS_and_MDS.md

**Comment by avatarneil** [comment]
@vperron  I agree, having just a boolean for `wheelchair_boarding` feels wrong, and also not super applicable for micro-mobility (unless there are handicap accessible scooters). How about an `accessibility_options` array that (for now) would just be empty? I can certainly imagine that this would be applicable for vehicles as well (especially for non-micro-mobility modes). 

Agreed having a pointer to `gbfs_station_id` could be beneficial, though I think that while it works well for the Provider API, it would be tricky for the Agency API, as for a given Stop I believe there could be multiple `gbfs_station_id`s for the same logical Stop (because Providers have their own GBFS feeds). This may be a non-issue for GBFS, however I know for GTFS there is certainly fragmentation. Even though this specific PR is for Provider, I'd really like to keep the Provider and Agency definitions of Stops aligned as much as possible. 

The `status` field I think could be beneficial, totally onboard with that! I'm thinking we could just lift the enum definition from #441. 

The `reported_at` field definitely makes sense, however I'd prefer to name it `last_updated` -- this aligns with the RT `/vehicles` endpoint in Provider.

I'd be okay with switching from using the mapped values to a more explicit array of docks/vehicles, however I'm not sure that's entirely necessary (and may have some downsides). Having the maps certainly makes deriving counts of certain vehicle types at the Stops faster (constant time lookup as opposed to having to iterate through the entire array), and explicitly encoding an array of vehicles (with essentially all of the information for the vehicle) within the stop seems like duplicating data that is already available via the `/vehicles` endpoint to me. My counter proposal is, instead of dumping the maps, what if we kept them and additionally included a `vehicle_ids` array which would contain pointers to the vehicles which are currently at the stop? This would then give all of the information needed to be able to derive specific details about the vehicles which are at the stop from the `/vehicles` endpoint, without duplicating in the vehicle data.

**Comment by thekaveman** [comment]
@avatarneil:

> The `reported_at` field definitely makes sense, however I'd prefer to name it `last_updated` -- this aligns with the RT `/vehicles` endpoint in Provider.

+1 on this approach (and the original idea to include this info). Does it make sense to keep this field at the top-level of the payload, to further align with `/vehicles`? I can see it making sense keeping it as a field of the stop, if different stops are updated at different frequencies.

We should also probably include `ttl` as well.

> The status field I think could be beneficial, totally onboard with that! I'm thinking we could just lift the enum definition from #441.

Agreed. #441 makes the note that some GBFS station status were purposely omitted (`is_installed`, `is_renting`, `is_returning`) do we want to include them all for completeness? It sounds like #441's concept of `open` includes `[is_renting, is_returning]`, and maybe it isn't necessary to be that explicit since `num_vehicles_available` and similar fields can be used to indicate if renting/returning is supported.

@vperron:

> a good chance to be integrated into the 1.0.0 milestone without too much hassle, especially if the changes to `/trips` and `/status_changes` are out of scope for 1.0 so far.

I would rather see this spec and the corresponding `/trips` and `/status_changes` updates go in for 1.0.0 at the same time. It's already going to be a major breaking release, we may as well take advantage.

Other feedback:

**`lat`/`lng`**: I'm a little torn. On one hand, every other Provider API uses a `*location` field with type GeoJSON Feature, so alignment with this could be beneficial (#441 takes this approach). On the other hand, GBFS [`station_information.json` uses explicit `lat`/`lon`](https://github.com/NABSA/gbfs/blob/master/gbfs.md#station_informationjson) fields, as does the [Agency telemetry APIs](https://github.com/openmobilityfoundation/mobility-data-specification/tree/master/agency#telemetry-data) (though in that case the fields are `lat`/`lng` as here).

I lean towards consistency within the Provider APIs.

**Comment by schnuerle** [comment]
Discussions from the call today.

Agreed to merge the different PRs into one joint Agency/Provider PR. @avatarneil volunteered to do that work.  Includes synthesis of ideas from Issues #374 #427 #438, PRs #403 #427 #442 #441.

Shared definitions should go in the general_information.md file (though in the future we may need a different place for this type of info).

People liked using geoJSON instead of simple lat/lng points since agency is already using this and it would provide a more general way to define areas like in #503. @Retzoh will leave more thought on this here.

Agree that a list of vehicle IDs by reference is better than returning vehicle info within this spec's payload.

Will not include the ideas of a history in stops.  That will be left for 1.1.0.

Will be marked as beta for 1.0.0, and beta label could be removed for 1.1.0 if consensus is reached and new additions are well defined.

Please leave any thoughts or comments you may have on this.

**Comment by avatarneil** [comment]
@schnuerle Re: Coalescing PRs into one -- how about I just merge in the Agency PR into this one, and we just use this as the main PR, as opposed to creating a new one? There's been a lot of meaningful discussion on this PR that I think would be buried a bit if we closed and cut a new one.

**Comment by schnuerle** [comment]
@avatarneil Yes I think that works just fine. We can rename.  You are right there is good discussion to keep front and center.

**Comment by avatarneil** [comment]
Anyone have thoughts around @dirkdk's suggestion of adding information about if Stops have charging capabilities? The more I think about it, if we're trying to include more information at the 'spot' level, it might be worth going with what @vperron suggested around having spots be in an array rather than in a map. I'm not a huge fan of composite keys personally, and having a map that's like: `vehicle_type: { propulsion_type: { charging/non_charging: number } } }` definitely seems like an anti-pattern. 

Additionally, if we moved towards spots being an array, we could have a unique identifier for the spots within a Stop! I could certainly see a use-case down the road where there are city managed Stops that a provider could reserve a particular spot at, and this would pave the way for that rather elegantly I think. 

If we don't make any progress on this by the Provider WG call this week, I'll bring it up :)

**Comment by thekaveman** [comment]
Conversation on 6/18:

* Agreed that GeoJSON is best representation for location information

* Agreed that this endpoint should mirror stop/station status concepts from GBFS, i.e. GBFS `is_installed`, `is_renting`, `is_returning` should be explicitly represented instead of a different set of status like `open`, `closed` a la #441.

* Time reporting should be matched with `/vehicles`: use `last_updated` and `ttl` at the top-level, introduce `last_reported` into the stop level.

**Comment by avatarneil** [comment]
Hey folks! I've started working on incorporating the changes discussed in the Provider Services WG meeting last week. As I'm building out the `Spot` data type, and am thinking more about `charging`, and I wonder if there's a need to describe the types of chargers in more detail. The question in my mind comes down to: is charging universal (I don't think it is), and if not, how do we represent the different types of charging options for a given `Spot`? Is there a standard that describes the different types of chargers that electric scooters/bicycles have in an enumerable way?

Worst case, I suppose we can just describe the different charges by some combination of vehicle_type & propulsion_type, but especially as we extend to new modes and have things like electric cars, different charging types will become very nuanced, so I think it's worth bringing this up now.

**Comment by karcass** [comment]
I suggest not including charging in the first release and giving it the consideration it merits in the 1.1 timeframe.

**Comment by avatarneil** [comment]
Still working on `Spots`, a couple more questions that arise:
1. Should we include the `device_ids` for vehicles currently at a Stop at the Spot level?
2. If we include `Spots`, and the `device_ids` live at that level, should we get rid of `capacity, num_spots_available, num_spots_disabled, num_vehicles_available, num_vehicles_disabled` from the Stop definition? It seems to me like they can all be derived based on the information for each Spot within a Stop (in combination with some queries to `/vehicles` to get information about the vehicles currently occupying the various Spots), however this would make it so Spots are required. 
3. There was discussion about having a particular Spot that could, say, fit 6 scooters, but only 2 bikes. Should we try to tackle this yet, or save that for a subsequent PR (maybe in the 1.1 timeframe)?

**Comment by avatarneil** [comment]
In light of trying to get this merged pretty soon, what do people think about holding off on the detailed `Spots` changes until 1.1? I think it merits a bit more consideration.

File: agency/README.md [hunk 1]
```diff
@@ -263,6 +272,35 @@ A standard point of vehicle telemetry. References to latitude and longitude impl
 | `gps.hdop`     | Float          | Required if Available | Horizontal GPS or GNSS accuracy value (see [hdop](https://support.esri.com/en/other-resources/gis-dictionary/term/358112bd-b61c-4081-9679-4fca9e3eb926)) |
 | `gps.satellites` | Integer      | Required if Available | Number of GPS or GNSS satellites
 | `charge`       | Float          | Required if Applicable | Percent battery charge of vehicle, expressed between 0 and 1 |
+| `stop_id`      | UUID           | Required if Applicable | Stop that the vehicle is currently located at. Only applicable for _docked_ Micromobility. See [Stops](../general-information.md#stops) |
+
+## Stops
+
+The `/stops` endpoint allows an agency to register Stops.
```
[the remark this answers is not in the document]
Comment (avatarneil) [reply agency/README.md:279]: When I first began working on the concept of Stops, it was initially designed such that a city would define the Stops that providers could use (e.g. could be used for designating virtual drop-off zones like the Venice Beach Special Operations Zone). That being said, I think you're right that we need a mechanism for providers registering stops with an agency via the Agency API as well I think. I'll reword so that it it something like:
```suggestion
The `/stops` endpoint allows an agency to register city-managed Stops, or a provider to register self-managed Stops.
```

[the remark this answers is not in the document]
Comment (avatarneil) [reply agency/README.md:279]: We probably ought to also extend it such that Stops in Agency have some encoding for who manages the stop. Perhaps just a `provider_id`?

[the remark this answers is not in the document]
Comment (Retzoh) [reply agency/README.md:279]: > When I first began working on the concept of Stops, it was initially designed such that a city would define the Stops that providers could use

👍 Cities don't need to use the API to create stops since they have access to the underlying implementation but it can't do any harm to mention they can.

> We probably ought to also extend it such that Stops in Agency have some encoding for who manages the stop. Perhaps just a `provider_id`?

I agree we should add a `provider_id`, also in case a provider manages a stop for another provider.

**Comment by avatarneil** [comment]
Poll for 'Spot' naming https://strawpoll.com/cbqpyog7o

**Comment by schnuerle** [comment]
@thekaveman thank you for the rebase!  

It looks like 'spaces' is the consensus term to replace 'spots' with 4 votes.  

@avatarneil can you rename all references to spots with spaces?

Let's also talk about the differences between MDS and GBFS here. 

1. station, docks, bikes **vs** stop, places, vehicles - I think these differences are ok, since MDS has a broader scope that GBFS and some of the GBFS naming may be a bit of a relic despite GBFS supporting more vehicle types.  Will GBFS change this wording eventually, or keep as is? I'm not sure. 
1. How does 'stop status' work in MDS?  Is it an array of the 3 options with a boolean for each?  It's not clear to me looking at the spec.
1. Is there a reason to keep MDS stops as one endpoint, or should it be broken out into `stop_status` and `stop_information`, like in GBFS?  I assume this is done to reduce the payload size and split up items that change less frequently.  With MDS it may make sense to keep as is since this info may be expected to change more frequently.
1. MDS is using a GeoJSON Point Feature for location, vs a lat/lon point, which makes sense to me for consistency within MDS.  However, should this be a geography instead of a point? I think we discussed this on calls.
1. Note MDS is introducing the idea of parent stops (a hierarchy) and adding vehicle_ids (which makes sense for authenticated MDS).

| GBFS station_status  |  MDS stops |
|---|---|
|  **station**_id  |  **stop**_id |
|  num_**docks**_available | num_**places**_available  |
| num_**docks**_disabled  | num_**places**_disabled  |
| num_**bikes**_available  |  num_**vehicles**_available |
| num_**bikes**_disabled  |  num_**vehicles**_disabled |
| is_installed | **stop status** is_installed |
| is_renting | **stop status** is_renting  |  
| is_returning | **stop status** is_returning |
| last_reported    | last_reported |

| GBFS station_information  |  MDS stops |
|---|---|
|  **station**_id  |  **stop**_id |
|  **lat** | **location**  |
| **lon**  | **location**  |
| name  |  **stop**_name |
| short_name | short_name |
| address | address  |  
| cross_street | cross_street |
| region_id    | region_id |
| post_code | post_code |
| rental_methods | rental_methods |
| capacity |  capacity  |
| X | **parent_stop** |
| X | **vehicle_ids** |

Thoughts on changes here are welcome, even though the release candidate will be made this Monday.  During the release approval process the OMF Tech Council and Board can discuss tweaks to this proposal, and comments made here will help influence those decisions.

**Comment by avatarneil** [comment]
> It looks like 'spaces' is the consensus term to replace 'spots' with 4 votes.

I'll replace 'spots' with 'spaces'.

> @avatarneil what happened to the 'Place' option, or did I just make that up?

'Place' totally slipped my mind, but I do remember that from the call -- don't think you made it up :)

> 2. How does 'stop status' work in MDS?  Is it an array of the 3 options with a boolean for each?  It's not clear to me looking at the spec.

It's an object with the properties, e.g.
```json
{
  "is_installed": true,
  "is_renting": false,
  "is_returning": true,
}
```

> 3. Is there a reason to keep MDS stops as one endpoint, or should it be broken out into `stop_status` and `stop_information`, like in GBFS?  I assume this is done to reduce the payload size and split up items that change less frequently.  With MDS it may make sense to keep as is since this info may be expected to change more frequently.

I think we ought to keep it as one endpoint (at least for now) -- this aligns well with how we report on `vehicles` in MDS.

> 4. MDS is using a GeoJSON Point Feature for location, vs a lat/lon point, which makes sense to me for consistency within MDS.  However, should this be a geography instead of a point? I think we discussed this on calls.

This confusion is one of the reasons why I was pushing for having simple `lat/lng` properties in the object, and using Geography (via a `geography_id`) for any GeoJSON. What we have now, is _both_ `location` represented as a GeoJSON Point Feature, as well as supporting a linkage to a Geography via a `geography_id` (allowing a full GeoJSON Feature Collection), which is a little odd. 

> 5. Should `stop_name` be changed to `name` to match GBFS and be more generic?

That sounds fine by me.

Cc. @schnuerle

**Comment by schnuerle** [comment]
Thank you for the updates and clarifications @avatarneil! 

I think this PR is ready for the 1.0.0 Release Candidate today.

For discussion during the review process:

1. Should we use 'place' instead of 'space'? Mostly because place starts with 'P' so would be easier to say along with with word 'Stops', and 'space' subtly references cars with 'parking space'.
1. We have not had a case of multiple IDs in an array like we do with the 'devices' field here.  Should these fields instead have '_ids' at the end, like GBFS does with 'station_ids' and 'region_ids'? Either way this is starting a naming standard for MDS.
1. Should we intentionally explain our naming of stop/spaces/vehicles, or should we instead align with GBFS station/docks/bikes even though the terminology is narrower?  We can ask MobilityData and GBFS community if there is any intention to broaden in the future.

File: agency/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Retzoh) [reply agency/README.md:239]: @avatarneil it looks like the wording update 

> The `/stops` endpoint allows an agency to register city-managed Stops, or a provider to register self-managed Stops.

got missed in the merging process, could you make a quick update including it?

File: agency/README.md [hunk 2]
```diff
@@ -231,17 +232,47 @@ A standard point of vehicle telemetry. References to latitude and longitude impl
 | `gps.hdop`     | Float          | Required if Available | Horizontal GPS or GNSS accuracy value (see [hdop][hdop]) |
 | `gps.satellites` | Integer      | Required if Available | Number of GPS or GNSS satellites
 | `charge`       | Float          | Required if Applicable | Percent battery charge of vehicle, expressed between 0 and 1 |
+| `stop_id`      | UUID           | Required if Applicable | Stop that the vehicle is currently located at. Only applicable for _docked_ Micromobility. See [Stops][stops] |
+
+## Stops
+
+The `/stops` endpoint allows an agency to register Stops.
```
Comment (Retzoh) [inline agency/README.md:239]: @avatarneil I think we forgot the wording update here, could you add a quick patch with it?

```suggestion
The `/stops` endpoint allows an agency to register city-managed Stops, or a provider to register self-managed Stops.
```

Comment (schnuerle) [reply agency/README.md:239]: I've made this update in the release cleanup PR here: https://github.com/openmobilityfoundation/mobility-data-specification/pull/539/commits/983a99fbb84d6d2ea805bcb82676a63bb431a605

File: agency/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (schnuerle) [reply agency/README.md:239]: Resolved here: https://github.com/openmobilityfoundation/mobility-data-specification/pull/539/commits/983a99fbb84d6d2ea805bcb82676a63bb431a605

**Comment by schnuerle** [comment]
Some updates from the [Provider Services call](https://github.com/openmobilityfoundation/mobility-data-specification/wiki/Web-Conference,-2020.07.02-(Provider-Services-WG)#agenda-and-minutes) today:

> 1. Should we use 'place' instead of 'space'? Mostly because place starts with 'P' so would be easier to say along with with word 'Stops', and 'space' subtly references cars with 'parking space'.

No opinion one way or the other.  I like "Place."

> 3. Should we intentionally explain our naming of stop/spaces/vehicles, or should we instead align with GBFS station/docks/bikes even though the terminology is narrower?  We can ask MobilityData and GBFS community if there is any intention to broaden in the future.

Heidi Guenin with MobilityData who runs GBFS said that there is a PR to change bikes to vehicles. Car sharing is a new use case. Bottom line is don't align to GBFS terminology right now.

**Comment by drtyh2o** [comment]
@schnuerle Have we considered `position`, that is, stops and stop positions?

**Comment by schnuerle** [comment]
> @schnuerle Have we considered `position`, that is, stops and stop positions?

Yes we did, it was in the [poll](https://github.com/openmobilityfoundation/mobility-data-specification/pull/427#issuecomment-650162856) we took (though Place was missing) and didn't get any votes.

Based on the feedback/votes on the City Services call today and the Provider Services call last week, I'm going to rename `Spaces` to `Places` in the release candidate PR #544.
